patches/server/0739-Add-CompostItemEvent-and-EntityCompostItemEvent.patch

```diff
@@ -0,0 +1,97 @@
++            BlockState iblockdata1 = ComposterBlock.addItem(iblockdata, DummyGeneratorAccess.INSTANCE, blockposition, itemstack, rand, entity); // Paper
+             if (iblockdata == iblockdata1 || org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, blockposition, iblockdata1).isCancelled()) {
+                 return iblockdata;
+             }
+-            iblockdata1 = ComposterBlock.addItem(iblockdata, (LevelAccessor) worldserver, blockposition, itemstack, rand);
++            // Paper start
++            iblockdata1 = ComposterBlock.addItem(iblockdata, (LevelAccessor) worldserver, blockposition, itemstack, rand, entity);
++            if (iblockdata1 == null) {
++                return iblockdata;
++            }
++            // Paper end
+             // CraftBukkit end
+ 
+             itemstack.shrink(1);
+@@ -294,13 +304,40 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
+         // CraftBukkit start
+         return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble());
+     }
++    // Paper start
++    static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item, @Nullable Entity entity) {
++        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble(), entity);
++    }
++    // Paper end
+ 
+     static BlockState addItem(BlockState iblockdata, LevelAccessor generatoraccess, BlockPos blockposition, ItemStack itemstack, double rand) {
++        // Paper start
++        return ComposterBlock.addItem(iblockdata, generatoraccess, blockposition, itemstack, rand, null);
++    }
++
++    static BlockState addItem(BlockState iblockdata, LevelAccessor generatoraccess, BlockPos blockposition, ItemStack itemstack, double rand, @Nullable Entity entity) {
++        // Paper end
+         // CraftBukkit end
+         int i = (Integer) iblockdata.getValue(ComposterBlock.LEVEL);
+         float f = ComposterBlock.COMPOSTABLES.getFloat(itemstack.getItem());
+ 
+-        if ((i != 0 || f <= 0.0F) && rand >= (double) f) {
++        // Paper start
++        io.papermc.paper.event.block.CompostItemEvent event = null;
++        if (generatoraccess != org.bukkit.craftbukkit.util.DummyGeneratorAccess.INSTANCE) {
++            boolean willRaiseLevel = !((i != 0 || f <= 0.0F) && rand >= (double) f);
++            if (entity == null) {
++                event = new io.papermc.paper.event.block.CompostItemEvent(org.bukkit.craftbukkit.block.CraftBlock.at(generatoraccess, blockposition), itemstack.getBukkitStack(), willRaiseLevel);
++            } else {
++                event = new io.papermc.paper.event.entity.EntityCompostItemEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(generatoraccess, blockposition), itemstack.getBukkitStack(), willRaiseLevel);
++            }
++        }
++
++        if (event != null && !event.callEvent()) {
++            return null;
++        }
```

Comment:
Can be moved inside the DummyGeneratorAccess.INSTANCE conditional check so you don't have to check if the event is null